jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/SecurityHandler.java

```diff
@@ -825,15 +840,149 @@ int compare(MappedResource<Constraint> c1, MappedResource<Constraint> c2)
+            super(handler);
+        }
+
+        /**
+         * <p>Associates the given {@link Constraint} to the given path and HTTP method.</p>
+         *
+         * @param pathSpec the {@link PathSpec} associated to the given constraint
+         * @param method the HTTP method associated to the given constraint, or {@code null}
+         * to indicate all HTTP methods
+         * @param constraint the constraint to associate
+         * @return the previous constraint associated with the given path and HTTP method,
+         * or {@code null} is there was no association
+         */
+        public Constraint put(String pathSpec, String method, Constraint constraint)
+        {
+            return put(PathSpec.from(pathSpec), method, constraint);
+        }
+
+        /**
+         * <p>Associates the given {@link Constraint} to the given path and HTTP method.</p>
+
+         * @param pathSpec the {@link PathSpec} associated to the given constraint
+         * @param method the HTTP method associated to the given constraint, or {@code null}
+         * to indicate all HTTP methods
+         * @param constraint the constraint to associate
+         * @return the previous constraint associated with the given path and HTTP method,
+         * or {@code null} is there was no association
+         */
+        public Constraint put(PathSpec pathSpec, String method, Constraint constraint)
+        {
+            Objects.requireNonNull(pathSpec);
+            if (method == null)
+                method = ALL_METHODS;
+            Objects.requireNonNull(constraint);
+            Map<String, Constraint> methodConstraints = _constraints.computeIfAbsent(pathSpec, k -> new HashMap<>());
+            Constraint result = methodConstraints.put(method, constraint);
+            if (result != null)
+                recomputeKnownRoles();
+            else
+                _knownRoles.addAll(constraint.getRoles());
+            return result;
+        }
+
+        /**
+         * <p>Associates the given {@link Constraint} to the given path and HTTP methods.</p>
+         *
+         * @param pathSpec the {@link PathSpec} associated to the given constraint
+         * @param methods the list of HTTP methods associated to the given constraint
+         * @param constraint the constraint to associate
+         */
+        public void put(PathSpec pathSpec, List<String> methods, Constraint constraint)
+        {
+            if (methods.isEmpty() || methods.contains(ALL_METHODS))
+                throw new IllegalArgumentException("Invalid method list");
+            methods.forEach(method -> put(pathSpec, method, constraint));
+        }
+
+        @Override
+        protected Constraint getConstraint(String pathInContext, Request request)
+        {
+            List<MappedResource<Map<String, Constraint>>> matches = _constraints.getMatches(pathInContext);
+
+            if (matches == null || matches.isEmpty())
+                return Constraint.ALLOWED;
+
```

Comment:
Is it worthwhile having a short cut for the common case of only a single match? Or in JIT do we trust?